### PR TITLE
Promotions remember their base code

### PR DESCRIPTION
**Pull request by jrochkind**
When multi-promotions were added to Solidus, they seem to have done so in a way that the base code is not stored on the promotion, it is purely a transitory property of the promotion creation process. 

This makes it hard to later create more promotion codes using the same base code. 

It also means when later looking at the Promotion in an admin edit screen, the first code from the db is shown on screen labelled as 'base code', rather than the base code itself. Which is kinda messy. 

This PR:
1. Adds a Promotion#base_code, and changes the controller to record the base code on #create. This was done in the _least changes_ way possible. So various solidus internal API is a little bit odd with this change. base_code should arguably be sent in params as an attribute of the promotion instead of a separate builder argument, and at several other parts of the API, etc. 
2. Changes the CodeBuilder so it can be used to add one more code and still get a suffix -- now it applies a suffix even if you're only asking for one code if there's existing codes. (Still no suffix if 1 code and no existing codes). 
3. Does _not_ change the backend UI at all (other than fixing base_code to display properly), including does NOT actually provide a backend UI to provide an 'add more codes' . this could be a future step. I think the promotion creation backend UI is kind of confusing for newcomers in general, and possibly needs more work beyond adding an 'add more codes' button.

### CODE REVIEW

**Comment by bbuchalter** [comment]
I personally think adding additional codes after a promotion is created is a _great_ feature and this PR moves us in that direction. ❤️

File: backend/app/controllers/spree/admin/promotions_controller.rb [hunk 1]
```diff
@@ -28,7 +26,7 @@ def create
 
       def load_bulk_code_information
         @promotion_builder = Spree::PromotionBuilder.new(
-          base_code: @promotion.codes.first.try!(:value),
+          base_code: @promotion.base_code || @promotion.codes.first.try!(:value),
```
Comment (bbuchalter) [inline backend/app/controllers/spree/admin/promotions_controller.rb]: If these two items are interchangeable, perhaps a migration to set that data on `base_code` would be helpful?

File: core/app/models/spree/promotion_code/code_builder.rb [hunk 2]
```diff
@@ -26,7 +26,7 @@ def build_promotion_codes
   private
 
   def codes
-    if num_codes > 1
+    if (num_codes + promotion.codes.size) > 1
```
Comment (bbuchalter) [inline core/app/models/spree/promotion_code/code_builder.rb]: It's a bit unclear to me why this change is required. A more detailed commit message would be welcome.

File: backend/app/controllers/spree/admin/promotions_controller.rb (hunk 1 above)
Comment (jrochkind) [inline backend/app/controllers/spree/admin/promotions_controller.rb]: They aren't exactly interchangeable, but there's no good way to write a migration to set that data, because that data simply isn't there -- that's what this PR fixes. 

I suppose you could write a migration that looked through all existing promotions, _if_ they have promotion codes and _if_ those codes all begin with the same prefix (assuming that prefix is always followed by an underscore, which the existing CodeBuilder does, but you could hypothetically have a custom CodeBuilder which does something else), then extract that prefix and set it as base_code?

That seemed far too error-prone and weird to try, to me.

Comment (jrochkind) [inline backend/app/controllers/spree/admin/promotions_controller.rb]: to be more clear: Prior to this PR, if you look at an existing Promotion, the _first_ promotion code (if any) is shown on screen labelled (labelled incorrectly) "base code". 

That's just a mess, but the actual base code was not stored anywhere to display instead. After this change, the actual base code is stored somewhere and can be displayed properly. But for backwards compatibility and with existing promotions that do not have a stored base code, it still falls through to the old behavior. 

As I warn in the PR, this change leaves a _lot_ of internal API (class/method) API being kind of confusing. A lot of existing design is... not the choices I would have made. But I intentionally made the smallest possible change to get a properly stored base_code. 

If people would rather I change more to make the internal API more reasonable, I certainly can -- but it's very challenging to figure out how to do so in a way that is both backwards compatible and makes the code less rather than more confusing.

File: core/app/models/spree/promotion_code/code_builder.rb (hunk 2 above)
Comment (jrochkind) [inline core/app/models/spree/promotion_code/code_builder.rb]: Prior to this change, if you tell the CodeBuilder to create one code, it _insists_ on creating that code _without_ a suffix -- the one code created is just the 'base_code' by itself. 

Even ignoring actual admin UI, I had tried to use the CodeBuilder itself to create _one_ _additional_ code, with suffix -- I could not do so, it won't do it. 

The CodeBuilder was I guess written with the assumption that it would only be used on new promotion creation. 

So this change keeps the intent and backwards compatible API of the CodeBuilder -- which seems to be "Decide whether to use a suffix based on how many codes there are" -- but makes it work for both about-to-be-created new promotions, and existing promotions. For existing promtions "how many codes there are" needs to take into account not just the codes you are about to add, but the codes that existed already as well. 

Again, this is not the API I would have done, but it's what's there, and I'm trying to not rewrite the whole thing, and maintain backwards compatibility. 

I am happy to ammend the commit message to be whatever you want if you have a suggestion.

Comment (jrochkind) [inline core/app/models/spree/promotion_code/code_builder.rb]: I am also happy to rewrite the internal API to be more consistent. If I didn't care about backwards compatibility, I wouldn't have the CodeBuilder deciding whether to use a suffix based on how many codes there are at all, I'd have another option. If I did care about backwards compatibility... it may have to do some kind of `*args` checking thing. 

Initially I was going down that route, but I think the chances of me getting such a PR approved are not very high, so didn't want to do all that work only to have it sit here forever without getting approved.

**Comment by jordan-brough** [comment]
@jrochkind thanks for raising this issue!

I've got a slightly different suggestion I'd like to offer on tackling this: What if we introduce a `PromotionCodeBatch` model?

I very much agree with all of these points:

> Prior to this PR, if you look at an existing Promotion, the first promotion code (if any) is shown on screen labelled (labelled incorrectly) "base code".
>
> That's just a mess, but the actual base code was not stored anywhere to display instead.

...

> hard to later create more promotion codes using the same base code.

...

> The CodeBuilder was I guess written with the assumption that it would only be used on new promotion creation.

...

> I wouldn't have the CodeBuilder deciding whether to use a suffix based on how many codes there are at all, I'd have another option.

Along with those, here are a few additional points of pain I've had with the current system:

- Multiple promotion codes on a single promotion should not be required to have the same base code, or even any base code at all. (Which actually argues against having a `spree_promotions.base_code` DB field)
- The admin can easily time out when trying to generate large batches of promotion codes (e.g. 100,000 codes)

I think we might be able to help with all these issues if we added a PromotionCodeBatch model and update the admin API a bit.

Specifics:

- PromotionCodeBatch Schema:
  - promotion_id
  - base_code
  - number_of_codes
  - status (processing/completed/errored)
- `PromotionCodeBatch` would not be not required.  You can add a code to a promotion without it being in a batch (makese sense, imo, and helps with backward-compatibility)
- `PromotionCodeBatch`es can be processed asynchronously. And since they're in the DB we can provide an admin UI on their status.
- Provide a different admin UI for generating single codes vs batches of codes and send them in separate controller parameters
- Don't use CodeBuilder for single codes
- Deprecate calling `CodeBuilder` with `num_codes == 1` instead of changing the behavior (and in the future make it do the same thing for `== 1` as it does for `> 1`)
- Deprecate `:base_code` as a parameter to PromotionBuilder instead of changing the behavior

Thoughts?  I know that's a much larger request, but personally I'd rather not cement the idea of `spree_promotions.base_code` as a data point, and this would be helpful to our store so I'd be happy to collaborate on it or try taking it on ourselves if people think it's an OK idea.

File: core/app/models/spree/promotion_builder.rb [hunk 3]
```diff
@@ -15,8 +15,10 @@ class Spree::PromotionBuilder
 
   # @param promotion_attrs [Hash] The desired attributes for the newly promotion
   # @param attributes [Hash] The desired attributes for this builder
-  # @param user [Spree::User] The user who triggered this promotion build
   def initialize(attributes = {}, promotion_attributes = {})
+    if base_code = attributes[:base_code]
+      promotion_attributes[:base_code] = base_code
```
Comment (jhawthorn) [inline core/app/models/spree/promotion_builder.rb]: This should avoid modifying the hash passed in.

**Comment by jrochkind** [comment]
Yep, I agree with @jhawthorn, this is basically meant to fix what I think was an omission in the implementation of the multi-code feature, it should have included this to begin with. Just fixing it in the simplest way possible, entirely backwards compatible.  I hope the slight cleanup 

I'll fix the mutation of the input parameter that @jhawthorn pointed out. 

If committers don't want to make this fix to the current architecture, insisting instead on a more major refactoring of the feature (perhaps backwards compat perhaps not), you can reject/close this PR.  I can't commit to doing that major refactoring myself though, especially cause I've found that major refactorings are unlikely to be approved by committers who will disagree with each other about how it should look. :) 

If this does get accepted, I think the next step is some slight fixes to the admin UI, to deal with confusion several people have reported on the slack channel about how to create a promotion with just _one_ base code.  With a base code stored, the UI can actually tell whether it's a multi-code promotion or a single-code promotion, and adjust labels accordingly to only say 'base code' if it really is a base code. 

Really, I think the original multi-code implementation was merged not really finished, just trying to finish it in a backwards compat way.

File: core/app/models/spree/promotion_builder.rb (hunk 3 above)
Comment (jrochkind) [inline core/app/models/spree/promotion_builder.rb]: done

**Comment by jordan-brough** [comment]
@vladstoick actually just implemented `PromotionCodeBatch` for us in our own store and we have been using it for the past week or so.  The aim was to upstream it.  Can you give us a week or so to see if we can get a PR up?  It could avoid the extra back-and-forth of changing this multiple times?  If it doesn't look good or if we don't manage to get it up then maybe we go forward with this PR?

**Comment by jrochkind** [comment]
Do what you gotta do. My use case that inspired this was I want to be able to say "Add N more codes" (where N can == 1!), and get codes added, without the code making this request having to know through some out of band source what the correct 'base code' is.  

I'd put handling that use case in as a request for your alternate proposed refactoring, and additionally a clear path for making the admin UI less confusing than it is now, although I don't think that has to be in the PR improving the behind-the-scenes. (Multiple reports of people not understanding how to add a single-code promotion from current admin UI).

**Comment by jhawthorn** [comment]
Closing now that #1524 has been merged
